### PR TITLE
Enhancements to Security Model and Application Configuration

### PR DESCRIPTION
**Pull request by Atabek04**
This PR focuses on significant improvements to Spring Security and overall application configuration, including:

- `Spring Security Enhancements`: Implemented JWT authentication, fixing related issues to ensure secure user access.
- `Account Locking Feature`: Introduced functionality to lock user accounts after multiple failed login attempts, enhancing security measures.
- `Spring Profiles & SQL Initialization`: Improved configuration management for seamless application setup.
- `Docker Configuration`: Added Docker support for streamlined deployment and testing.
- `Health Check Endpoint`: Integrated health check functionality for monitoring application status.

### CODE REVIEW

File: src/main/java/com/epam/gym/exception/JWTException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.epam.gym.exception;
+
+public class JWTException extends RuntimeException{
```
Comment (RYelzhan) [inline src/main/java/com/epam/gym/exception/JWTException.java:3]: Ctrl+alt+l

Comment (RYelzhan) [reply src/main/java/com/epam/gym/exception/JWTException.java:3]: Maybe in other custom exceptions as well.

File: src/main/java/com/epam/gym/security/jwt/JwtRequestFilter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+@Slf4j
+public class JwtRequestFilter extends OncePerRequestFilter {
+
+    private final JwtUtil jwtUtil;
+
+    public JwtRequestFilter(JwtUtil jwtUtil) {
```
Comment (RYelzhan) [inline src/main/java/com/epam/gym/security/jwt/JwtRequestFilter.java:30]: Consider @RequiredArgsConstructor

File: src/main/java/com/epam/gym/security/jwt/JwtUtil.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+@Slf4j
+public class JwtUtil {
+    @Value("${security.jwt.secret-key}")
+    private String secret;
+
+    @Value("${security.jwt.issuer}")
+    private String issuer;
+
+    @Value("${security.jwt.expiration-ms}")
+    private long expirationTime = 1000 * 60 * 60 * 10L;
```
Comment (RYelzhan) [inline src/main/java/com/epam/gym/security/jwt/JwtUtil.java:28]: Do not quite understand why you initialise it using @value and manually at the same time. Consider leaving one.

**Review by RYelzhan — APPROVED** [review]
Great work. I only have a little not mandatory suggestions, consider them.

File: src/main/java/com/epam/gym/actuator/DatabaseHealthIndicator.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        details.put("trainer_table", getTableRowCount("trainer"));
+        details.put("training_table", getTableRowCount("training"));
+        details.put("training_type_table", getTableRowCount("training_type"));
+
+        return Health.up().withDetails(details).build();
+    }
+
+    private Integer getTableRowCount(String tableName) {
+        try {
+            return jdbcTemplate.queryForObject("SELECT COUNT(*) FROM " + tableName, Integer.class);
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/actuator/DatabaseHealthIndicator.java:33]: Overall, there're doing the same thing, non of them will prevent from SQL injection anyway. But thanks for giving another option!

File: src/main/java/com/epam/gym/config/WebConfig.java [hunk 5]
```diff
@@ -18,9 +16,21 @@
...
+
+        registry.addMapping("/v1/**")
+                .allowedOrigins("http://localhost:8090")
+                .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS")
+                .allowCredentials(true);
+    }
+
     @Bean
     public MappingJackson2HttpMessageConverter jsonConverter() {
         return new MappingJackson2HttpMessageConverter();
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/config/WebConfig.java:36]: This CORS configuration for explicitly allowing specific HTTP methods from the specified origin.

File: src/main/java/com/epam/gym/controller/AuthController.java [hunk 6]
```diff
@@ -2,30 +2,53 @@
...
 @RequestMapping("/v1/auth")
 @Slf4j
 public class AuthController {
-    private final UserService userService;
+    private final AuthService authService;
 
     @PostMapping("/login")
-    public void login(@Valid @RequestBody UserCredentials credentials) {
-        userService.findByUsernameAndPassword(credentials.username(), credentials.password());
+    public ResponseEntity<Map<String, String>> login(@Valid @RequestBody UserCredentials credentials) {
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/controller/AuthController.java:28]: Sure! Declarative approach is smth new for me. I still not used to that)

File: src/main/java/com/epam/gym/exception/JWTException.java (hunk 1 above)
Comment (Atabek04) [reply src/main/java/com/epam/gym/exception/JWTException.java:3]: Well mentioned! Also I should make habit of formatting my code

File: src/main/java/com/epam/gym/security/jwt/JwtUtil.java (hunk 3 above)
Comment (Atabek04) [reply src/main/java/com/epam/gym/security/jwt/JwtUtil.java:28]: I guess I forgot to remove that part.

File: src/main/java/com/epam/gym/controller/AuthController.java [hunk 7]
```diff
@@ -2,30 +2,53 @@
...
+        return ResponseEntity.ok(authService.login(credentials));
+    }
+
+    @PostMapping("/refresh-token")
+    public ResponseEntity<Map<String, String>> refreshToken(@RequestBody Map<String, String> tokenRequest) {
+        String refreshToken = tokenRequest.get("refreshToken");
```
[the remark this answers is not in the document]
Comment (Atabek04) [reply src/main/java/com/epam/gym/controller/AuthController.java:34]: Couldn't really understand, to make constant of what ? Do you mean Map keys ?

[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/gym/controller/AuthController.java:34]: Yeah, map key of "refreshToken"

File: src/main/java/com/epam/gym/config/WebConfig.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/gym/config/WebConfig.java:36]: I meant @Bean public MappingJackson2HttpMessageConverter jsonConverter() { ... }. It is a default one spring uses under the hood.

File: src/main/java/com/epam/gym/controller/AuthController.java [hunk 8]
```diff
@@ -2,30 +2,53 @@
...
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
+import java.util.Map;
+
 @RestController
 @RequiredArgsConstructor
 @RequestMapping("/v1/auth")
 @Slf4j
 public class AuthController {
```
Comment (RYelzhan) [inline src/main/java/com/epam/gym/controller/AuthController.java:24]: Also, for your controllers. You seem to forgot about Swagger annotations for OpenApi documentation. consider adding them. Namely, I would consider @Operation, @ApiResponse (probably will need several one of them), and @SecurityRequirement. In advance, you can write them in interfaces which your controllers will implement. In this case, your buseness logic will be separated from documentation logic.

File: src/main/java/com/epam/gym/security/jwt/JwtUtil.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.stereotype.Component;
+
+import javax.crypto.SecretKey;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+@Slf4j
+public class JwtUtil {
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/security/jwt/JwtUtil.java:20]: Should we consider the `@UtilityClass` annotation from the Lombok library?

File: src/main/java/com/epam/gym/service/AuthService.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                user.setAccountNonLocked(false);
+                user.setLockoutTime(LocalDateTime.now());
+            }
+            userService.create(user);
+
+            Map<String, String> errorResponse = new HashMap<>();
+            errorResponse.put("error", "Invalid credentials. Attempts: " + attempts);
+            return errorResponse;
+        }
+    }
```
Comment (maximelyashevich) [inline src/main/java/com/epam/gym/service/AuthService.java:77]: Can you please decrease the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of this method?

File: src/main/resources/application-dev.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
+spring:
+  datasource:
+    url: jdbc:h2:file:./data/devdb;DB_CLOSE_ON_EXIT=FALSE;AUTO_RECONNECT=TRUE
+    username: sa
+    password: password
```
Comment (maximelyashevich) [inline src/main/resources/application-dev.yml:5]: Can you please remove such sensitive details and consider a more secure solution?

File: src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java [hunk 12]
```diff
@@ -53,7 +53,7 @@
 
 @ActiveProfiles("test")
 @ExtendWith({SpringExtension.class, TraineeServiceParameterResolver.class})
-@ContextConfiguration(classes = {ApplicationConfig.class})
+@ContextConfiguration(classes = {GymApp.class})
```
Comment (maximelyashevich) [inline src/test/java/com/epam/gym/api/TraineeControllerIntegrationTest.java:56]: Didn't really see your security tests. Please consider some useful articles on how to test security in Spring Boot applications:

- [Testing Method Security](https://docs.spring.io/spring-security/reference/servlet/test/method.html) (general overview)
- [Spring Security Testing](https://docs.spring.io/spring-security/site/docs/5.2.x/reference/html/test.html) (search by JWT)
- [Spring Security: Integration Tests](https://www.baeldung.com/spring-security-integration-tests) (practical examples)
- [Testing Security Configuration in Spring Boot](https://medium.com/@nagarjun_nagesh/testing-security-configuration-in-spring-boot-444b854bbd19) (practical examples)

The above materials are mostly for self-development.

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side, good job!

**Pull request merged**